torch/_export/serde/serialize.py

```diff
@@ -633,15 +632,150 @@ def serialize(self, exported_program: ep.ExportedProgram) -> Tuple[ExportedProgr
         )
 
 
+lib = Library("aten", "FRAGMENT")
+impl_lib = Library("aten", "IMPL")
+
+class GraphModuleOpUpgrader:
+    """Given model op set version number as well as compiler op set version number, do the following:
+    1. retrieve upgraders from somewhere (TorchScript API or new API) and pass it into this upgrader.
+    2. parse it and reorder for upgrading purpose.
+    3. register old versions of operators as custom ops.
+    4. prepare upgrader passes.
+
+    In `upgrade()` API run these upgrader passes.
+
+    An example of op_upgraders:
+    {
+        "aten::div__Scalar_0_3": (                              # versioned op name
+            "div._Scalar(self: Tensor, other: Scalar)",         # old schema
+            "
+def div__Scalar_0_3(self: Tensor, other: number) -> Tensor:     # upgrader in literal string
+  if (self.is_floating_point() or isinstance(other, float)):
+    return self.true_divide_(other)
+  return self.divide_(other, rounding_mode='trunc')
+",
+        ),
+    },
+    """
+
+    class UpgraderPass(ExportPassBase):
+        def __init__(self, old_target: Target, new_target: Target):
+            super().__init__()
+            self.old_target = old_target
+            self.new_target = new_target
+
+        def call_operator(
+                self,
+                op,
+                args: Tuple[torch.fx.node.Argument, ...],
+                kwargs: Dict[str, torch.fx.node.Argument],
+                meta: NodeMetadata,
+        ) -> ProxyValue:
+            if op == self.old_target:
+                return super().call_operator(self.new_target, args, kwargs, meta)
+            return super().call_operator(op, args, kwargs, meta)
+
+    def __init__(
+            self,
+            compiler_opset_version: Optional[Dict[str, int]] = None,
+            model_opset_version: Optional[Dict[str, int]] = None,
+            op_upgraders: Optional[Dict[str, Tuple[str, str]]] = None,
+            # TODO(larryliu): can add a new TS API: torch._C._get_upgraders_entry_map()
+    ):
+        self.compiler_opset_version = compiler_opset_version if compiler_opset_version else {}
+        self.model_opset_version = model_opset_version if model_opset_version else {}
+        # key is version number, value is a list of upgraders, keyed on op name
+        self.upgraders: List[Tuple[str, str]] = self._parse_upgraders(op_upgraders)
```

Comment:
does this need to be an attribute? We can directly pass this to _populate_passes()